routing/cross_mwm_index_graph.cpp

```diff
@@ -41,47 +80,112 @@ bool CrossMwmIndexGraph::IsTransition(Segment const & s, bool isOutgoing)
+        ("Mwm ", s.GetMwmId(), "has not been downloaded. s:", s, ". isOutgoing", isOutgoing));
 
   if (isOutgoing)
     return it->second.m_outgoing.count(s) != 0;
   return it->second.m_ingoing.count(s) != 0;
 }
 
-void CrossMwmIndexGraph::GetTwins(Segment const & /* s */, std::vector<Segment> & /* twins */) const
+void CrossMwmIndexGraph::GetTwins(Segment const & s, bool isOutgoing, std::vector<Segment> & twins)
 {
+  if (!IsTransition(s, isOutgoing))
+    CHECK(false, ("IsTransition(", s, ",", isOutgoing, ") returns false."));
   // @TODO(bykoianko) It's necessary to check if mwm of |s| contains an A* cross mwm section
   // and if so to use it. If not, osrm cross mwm sections should be used.
-  NOTIMPLEMENTED();
+  auto const segMwmMapping = GetRoutingMapping(s.GetMwmId());
+  CHECK(segMwmMapping, ("Mwm", s.GetMwmId(), "has not been downloaded. s:", s, ". isOutgoing", isOutgoing));
+
+  twins.clear();
+  vector<string> const & neighboringMwm = segMwmMapping->m_crossContext.GetNeighboringMwmList();
+
+  for (string const & name : neighboringMwm)
+    InsertWholeMwmTransitionSegments(m_numMwmIds->GetId(CountryFile(name)));
+
+  auto it = m_transitionCache.find(s.GetMwmId());
+  CHECK(it != m_transitionCache.cend(), ());
+
+  ms::LatLon const & latLog = isOutgoing ? GetLatLon(it->second.m_outgoing, s)
+                                         : GetLatLon(it->second.m_ingoing, s);
+  for (string const & name : neighboringMwm)
+  {
+    NumMwmId const numMwmId = m_numMwmIds->GetId(CountryFile(name));
+    TRoutingMappingPtr mappingPtr = GetRoutingMapping(numMwmId);
+    if (mappingPtr == nullptr)
+      continue;  // mwm was not loaded.
+
+    if (isOutgoing)
+    {
+      mappingPtr->m_crossContext.ForEachIngoingNodeNearPoint(latLog, [&](IngoingCrossNode const & node){
+        AddTransitionSegment(mappingPtr->m_segMapping, node.m_nodeId, numMwmId, twins);
+      });
+    }
+    else
+    {
+      mappingPtr->m_crossContext.ForEachOutgoingNodeNearPoint(latLog, [&](OutgoingCrossNode const & node){
+        AddTransitionSegment(mappingPtr->m_segMapping, node.m_nodeId, numMwmId, twins);
+      });
+    }
+  }
+
+  my::SortUnique(twins);
 }
 
 void CrossMwmIndexGraph::GetEdgeList(Segment const & /* s */,
-                                         bool /* isOutgoing */, std::vector<SegmentEdge> & /* edges */) const
+                                     bool /* isOutgoing */, std::vector<SegmentEdge> & /* edges */) const
 {
   // @TODO(bykoianko) It's necessary to check if mwm of |s| contains an A* cross mwm section
   // and if so to use it. If not, osrm cross mwm sections should be used.
   NOTIMPLEMENTED();
 }
+
+TRoutingMappingPtr CrossMwmIndexGraph::GetRoutingMapping(NumMwmId numMwmId)
+{
+  CountryFile const & countryFile = m_numMwmIds->GetFile(numMwmId);
+  TRoutingMappingPtr mappingPtr = m_indexManager.GetMappingByName(countryFile.GetName());
+  CHECK(mappingPtr, ("No routing mapping file for countryFile:", countryFile));
```

Comment:
```c++
TRoutingMappingPtr CrossMwmIndexGraph::GetRoutingMapping(NumMwmId numMwmId)
{
  CountryFile const & countryFile = m_numMwmIds->GetFile(numMwmId);
  TRoutingMappingPtr mappingPtr = m_indexManager.GetMappingByName(countryFile.GetName());
  CHECK(mappingPtr, ("No routing mapping file for countryFile:", countryFile));
  MappingGuard mappingPtrGuard(mappingPtr);

  if (!mappingPtr->IsValid())
    return nullptr; // mwm was not loaded.
  mappingPtr->LoadCrossContext();
  return mappingPtr;
}
```
If |numMwmId| is correct m_numMwmIds->GetFile(numMwmId) should return a correct |mappingPtr| and than m_indexManager.GetMappingByName(...) should return a |mappingPtr| which is not nullptr. I guess it should go this way. If not I prefer to be aware and investigate this case. Yes, it's a possible work around to return nullptr with LOG(ERROR, ...) but I prefer to leave CHECK to be aware it happens anyway (even in release).